File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/AppSettingsPage.java
package org.wordpress.android.e2e.pages;

import androidx.test.espresso.ViewInteraction;

import junit.framework.AssertionFailedError;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.hasSibling;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withContentDescription;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.allOf;

public class AppSettingsPage {
    private static ViewInteraction appSettingsLabel = onView(withText("App Settings"));
    private static ViewInteraction backArrow = onView(withContentDescription("Navigate up"));

    private static ViewInteraction editorTypeButton = onView(allOf(withId(android.R.id.summary),
            hasSibling(withText("Set editor type"))));

    public AppSettingsPage() {
        appSettingsLabel.check(matches(isDisplayed()));
    }

    public AppSettingsPage setEditor(String editorType) {
        try {
            editorTypeButton.check(matches(withText(editorType)));
        } catch (AssertionFailedError e) {
            editorTypeButton.perform(click());
            onView(allOf(withId(android.R.id.text1), withText(editorType))).perform(click());
        }

        return this;
    }

    public void goBack() {
        backArrow.perform(click());
    }
}

File: WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/AppSettingsPage.kt
package org.wordpress.android.e2e.pages

import android.R
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import junit.framework.AssertionFailedError
import org.hamcrest.Matchers

class AppSettingsPage {
    init {
        appSettingsLabel.check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }

    fun setEditor(editorType: String?): AppSettingsPage {
        try {
            editorTypeButton.check(ViewAssertions.matches(ViewMatchers.withText(editorType)))
        } catch (e: AssertionFailedError) {
            editorTypeButton.perform(ViewActions.click())
            Espresso.onView(
                Matchers.allOf(
                    ViewMatchers.withId(R.id.text1),
                    ViewMatchers.withText(editorType)
                )
            ).perform(ViewActions.click())
        }
        return this
    }

    fun goBack() {
        backArrow.perform(ViewActions.click())
    }

    companion object {
        private val appSettingsLabel = Espresso.onView(ViewMatchers.withText("App Settings"))
        private val backArrow = Espresso.onView(ViewMatchers.withContentDescription("Navigate up"))
        private val editorTypeButton = Espresso.onView(
            Matchers.allOf(
                ViewMatchers.withId(R.id.summary),
                ViewMatchers.hasSibling(ViewMatchers.withText("Set editor type"))
            )
        )
    }
}
